docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+  HVPA's policy for early horizontal scale-out.
+- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions.
+- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi.
+- Reduce Gardener custom code base and ongoing maintenance effort.
+
+### Non-Goals
+- There are other components currently scaled via HVPA. Changing the autoscaling mode for these is outside the scope
+  of this GEP.
+- Changing the autoscaling of `kube-apiserver`s other than the one in the shoot control plane, is outside the scope of
+  this GEP. These `kube-apiserver`s are subject to a different mode of autoscaling (e.g. vertical-only, where HVPA is
+  only used to stabilize VPA behavior), and it is advantageous from project execution perspective to address these as a 
+  separate concern.
+
+## Proposal
+ShootKapi HVPA is replaced by individual HPA and VPA. Undesirable interference between HPA and VPA is avoided through
+the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
+driven by CPU and memory usage metrics.
+
+### Rationale
+A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
+reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
+and stability (requests not too low).
+
+### Design Outline
+One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+An API request rate metric is deployed to shoot prometheus (one per shoot). It gets aggregated by the aggregate
+prometheus of the respective seed (one per seed) through Gardener's existing metrics aggregation mechanism.
+
+A new component named prometheus-metrics-adapter is deployed on each seed. It sources performance metrics data from the
+seed aggregate prometheus and makes it available in the K8s metrics format. The prometheus-metrics-adapter is registered
+as an extension API service to the seed's kube-apiserver and takes responsibility of providing all custom metrics for
+that kube-apiserver.
+
+The request rate signal driving HPA originates in the shoot prometheus. It is propagated to the aggregate seed
+prometheus, and then further mediated by the prometheus-metrics-adapter and exposed as custom metric in the seed
+kube-apiserver. The custom metric in turn is directly consumed by the seed's HPA controller.
+
+![Design outline](./assets/gep22-design-outline.png "Design outline")
+
+_Fig 2: Proposed design_
+
+### Element: New Metric in Shoot Prometheus
+A new `shoot:apiserver_request_total:sum` recording rule is added to the shoot prometheus. To minimize latency, it
```

Comment:
Thanks for pointing this out. I believe I'd be better off using a ":rate_by_pod" suffix, instead of ":sum". The following is the change I made, and also answers your question:

### Element: New Metric in Shoot Prometheus
A new `shoot:apiserver_request_total:rate_by_pod` recording rule is added to the shoot prometheus. It is the rate of API requests per second, broken down by ShootKapi pod. To minimize latency, it calculates request rate across the minimum period for which two distinct samples are reliably available.

I'll double check the new name against convention.